builtin/readline_osh.py

```diff
@@ -49,46 +48,67 @@ def __exit__(self, type, value, traceback):
 class BindXCallback(object):
     """A callable we pass to readline for executing shell commands."""
 
-    def __init__(self, eval):
-        # type: (Eval) -> None
+    def __init__(self, eval, mem, errfmt):
+        # type: (Eval, Mem, ui.ErrorFormatter) -> None
         self.eval = eval
+        self.mem = mem
+        self.errfmt = errfmt
 
     def __call__(self, cmd, line_buffer, point):
-        # type: (str, str, int) -> Tuple[int, str, str]
+        # type: (str, str, int) -> Tuple[int, str, int]
         """Execute a shell command through the evaluator.
 
         Args:
           cmd: The shell command to execute
           line_buffer: The current line buffer
           point: The current cursor position
         """
-        print("Setting READLINE_LINE to: %s" % line_buffer)
-        print("Setting READLINE_POINT to: %s" % point)
-        print("Executing cmd: %s" % cmd)
 
-        # TODO: add READLINE_* env vars later
-        # assert isinstance(line_buffer, str)
-        # self.mem.SetNamed(location.LName("READLINE_LINE"),
-        #                  value.Str(line_buffer),
-        #                  scope_e.GlobalOnly,
-        #                  flags=SetExport)
+        # Set READLINE_* env vars so they're available in case the command uses them
+        state.ExportGlobalString(self.mem, "READLINE_LINE", line_buffer)
+        state.ExportGlobalString(self.mem, "READLINE_POINT", str(point))
 
         # TODO: refactor out shared code from Eval, cache parse tree?
 
         cmd_val = cmd_eval.MakeBuiltinArgv([cmd])
         status = self.eval.Run(cmd_val)
 
-        return (status, line_buffer, str(point))
+        # Retrieve READLINE_* env vars and compare for changes
+        readline_line = self._get_rl_env_var('READLINE_LINE')
+        readline_point = self._get_rl_env_var('READLINE_POINT')
+        post_line_buffer = readline_line if readline_line is not None else line_buffer
+        post_point = int(
+            readline_point) if readline_point is not None else point
+
+        self.mem.Unset(location.LName('READLINE_LINE'), scope_e.GlobalOnly)
```

Comment:
I think we should use a context manager for unsetting to be "exception safe".    e.g. `eval.Run` may  raise an exception, and then this line won't be run

We have an idiom for this - `with ctx_Foo(...)`

https://oils.pub/release/latest/doc/oils-repo/mycpp/README.html#python-context-manager-c-constructor-and-destructor-raii

You can see many examples in say `core/state.py` -- the `__init__` pushes the state, and the `__exit__` pops it.  (And for our purposes, the `__enter__` must be there, but is ignored)

Reply:
I'm adding it as `ctx_EnvVars`. It's in `readline_osh.py`, but it's general enough that it could be used elsewhere, if there's other places osh needs to temporarily set/unset env vars.